scripts/concurrent_task_utils.py

```diff
@@ -45,24 +45,19 @@ def log(message, show_time=False):
 class TaskThread(threading.Thread):
     """Runs a task in its own thread."""
 
-    def __init__(self, func, verbose, semaphore, name):
+    def __init__(self, func, semaphore, name):
         super(TaskThread, self).__init__()
         self.func = func
         self.output = None
         self.exception = None
         self.stacktrace = None
-        self.verbose = verbose
         self.name = name
         self.semaphore = semaphore
         self.finished = False
 
     def run(self):
         try:
             self.output = self.func()
-            if self.verbose:
-                log('LOG %s:' % self.name, show_time=True)
-                log(self.output)
```

Comment:
Why are you removing this? This is a general logger for concurrent tasks, see if you can fix this without changing this?

Reply:
We don't need this now, because we are printing where we are running it, I tried to fix this by checking if we are getting a list of messages and ignore if we are getting a list but @seanlip disapproved that solution as that was not conceptually right(Sorry, I do not have the proof now, as the comment was resolved.)

Reply:
>  ignore if we are getting a list but @seanlip disapproved that solution as that was not conceptually right

Yeah, type checking is really not a good thing to do here!

I had a discussion with @bansalnitish on this today and we though this:

1. Have a function in linter class which will return a list of the linter functions.
2. In the pre_commit linter, file create a full list of these returned linter functions
3. Create a TaskThread object for each of these linter functions.
4. Run all the tasks in parallel.

(This way all the linter methods won't have to manage logging!)

Note: We have to find a way to run run "js_ts validator, parser" before running the lint test (not sure!)

You can discuss this with @bansalnitish in detail.

Reply:
Hi @Anshul,

I think what Sandeep's said is fine (I know it would not be easy and would take some time to implement such stuff).

Let me explain the problem with the example. We will consider the `check _check_non_test_files` mentioned here: https://github.com/oppia/oppia/blob/develop/scripts/linters/python_linter.py#L108.

In my opinion, I see the following problem with this:

1. We have self.verbose and then the print statement saying this check is started. We would have to write the same three four lines while we write any other check. 
I would like to see this being removed from each of the check and added it somewhere in common (basically somehow use concurrent utils).

2. We see two `PRINT` statement in this check -- once while the error is caught and once during the end of the check. Do we really want these PRINT statements? In addition, we are returning the summary_messages which we will be printed at the last + message list would also be printed for checks in these functions while tasks are created via concurrent_utils. So, in total an error in this message would be seen four times. Do you see this as a problem?
I would like to have two messages being printed out of this test. Once while the entire list of messages is printed at the end by passing summary_messages (we won't touch that flow). Once, we could list the errors that this check caught at the end of this check. The check should have not have any PRINT statement.


This is how I think could be done (we can discuss in detail and this will require some design analysis):

1. We can remove all the PRINT statements from the check
2. Create a task thread of all functions. We can change this line: https://github.com/oppia/oppia/blob/develop/scripts/linters/python_linter.py#L599 to return all functions basically change it to

```
return custom_linter.get_linter_functions(), third_party_linter.get_linter_functions
```

3. Accumulate all these function in pre_commit_linter or wherever suitable. Execute them parallelly using concurrent utils, which we print the output of each test once it is completed (In addition, we can convert these private check functions to public and we will be able to execute all the checks in parallel rather than executing classes in parallel -- I might be wrong because I don't understand why these functions are marked private).

This would fix both the problems.

So, two questions for you:
1. Do you see removing all print statements which are scattered in each check, do not help much in debugging, addition of 4 lines at the start to check verbose condition etc is a problem (as mentioned in points 1-2 above) and should be solved (let us not consider how it would be implemented)?
2. Do you think we can achieve the approach that I mentioned?

Tagging @seanlip @DubeySandeep too for there thoughts!

Reply:
I agree with the printing part, but running every method in parallel, I don't think this is a good idea to run each method in parallel as most of the methods depends on each other in class.

Reply:
This was discussed in the most recent meeting with @bansalnitish and @Hudda.

Could one of you please add a clear, full summary of the plan of action and the results of the discussion, so that @DubeySandeep can take a look and comment?

Reply:
OK, I can take this up.

As per our meeting:
1. We all agreed that the problems I posted above are genuine.
2. We decided to eradicate all the PRINT statements from the checks.
3. The checks would remain private and hence classes would be executed in parallel.
4. There would be no verbose statement too at the start of each check, it would be covered by concurrent_utils.
5. Each check would log two times, once while the checks run via concurrent_utils and once when the full list of summary_messages is printed at the end.
6. The API exposed by concurrent_utils should be such that it takes type from both the backend tests and the linter. Currently, the linter passed a list of string and backend tests pass strings. We will make this consistent so that both of them pass a similar object(it could be anything). 
7. If time allows (or may be in next milestone), we will even convert the summary_message that is returned from each check to a domain object with properties like check_name, status, errors, etc.

@Hudda could you please add if I missed something? 

Also, @DubeySandeep does this sounds good to you?

Reply:
@DubeySandeep ping!

Reply:
> Also, @DubeySandeep does this sounds good to you?


This sounds awesome! Thanks for discussing this with others and finalizing it! :)

Reply:
> The checks would remain private and hence classes would be executed in parallel.

We had planned to run each linter function in parallel, right?

Reply:
Yes, I agree. Though, this is more of design decision I would say. 

Anshul say's its fine to use classes in parallel. I'm fine with both to be honest, but I'm more inclined towards running functions in parallel. @Hudda can you please enlist why you don't want the functions to be converted to public?

Thanks!